File: 01-Programming-Languages/Kotlin/Data-Structures/LinkedList/merge_two_sorted_lists.kt
/*
 * Problem: Merge two sorted linked lists into one sorted list.
 * Approach: Iterative — compare nodes and link the smaller one.
 * Time Complexity: O(n + m)
 * Space Complexity: O(1)
 * Example: Input: 1 -> 3 -> 5, 2 -> 4 -> 6 -> Output: 1 -> 2 -> 3 -> 4 -> 5 -> 6 -> null
 */

class ListNode(var `val`: Int) {
    var next: ListNode? = null
}

fun mergeTwoLists(l1: ListNode?, l2: ListNode?): ListNode? {
    val dummy = ListNode(0)
    var tail: ListNode? = dummy
    var a = l1; var b = l2
    while (a != null && b != null) {
        if (a.`val` <= b.`val`) {
            tail!!.next = a; a = a.next
        } else {
            tail!!.next = b; b = b.next
        }
        tail = tail!!.next
    }
    tail!!.next = a ?: b
    return dummy.next
}

fun main() {
    val l1 = ListNode(1); l1.next = ListNode(3); l1.next!!.next = ListNode(5)
    val l2 = ListNode(2); l2.next = ListNode(4); l2.next!!.next = ListNode(6)
    var merged = mergeTwoLists(l1, l2)
    while (merged != null) {
        print("${merged.`val`} ")
        merged = merged.next
    }
}

File: 01-Programming-Languages/PHP/Data-Structures/LinkedList/merge_two_sorted_lists.php
<?php

/*
Problem: Merge Two Sorted Lists
Description: Merge two sorted linked lists into one sorted linked list.

Approach:
- Use a dummy head node for easy result building
- Compare heads of both lists, append the smaller one
- Advance pointer in the list whose node was taken
- Append remaining nodes of the non-empty list

Time Complexity: O(n + m)
Space Complexity: O(1)

Example:
Input:  list1 = 1 -> 2 -> 4, list2 = 1 -> 3 -> 4
Output: 1 -> 1 -> 2 -> 3 -> 4 -> 4
*/

class ListNode {
    public $val;
    public $next;

    public function __construct($val = 0, $next = null) {
        $this->val = $val;
        $this->next = $next;
    }
}

function mergeTwoLists(?ListNode $l1, ?ListNode $l2): ?ListNode {
    $dummy = new ListNode();
    $tail = $dummy;
    while ($l1 !== null && $l2 !== null) {
        if ($l1->val <= $l2->val) {
            $tail->next = $l1;
            $l1 = $l1->next;
        } else {
            $tail->next = $l2;
            $l2 = $l2->next;
        }
        $tail = $tail->next;
    }
    $tail->next = $l1 ?? $l2;
    return $dummy->next;
}

function printList(?ListNode $head): void {
    $vals = [];
    while ($head !== null) {
        $vals[] = $head->val;
        $head = $head->next;
    }
    echo implode(" -> ", $vals) . "\n";
}

$l1 = new ListNode(1, new ListNode(2, new ListNode(4)));
$l2 = new ListNode(1, new ListNode(3, new ListNode(4)));
echo "Merged: ";
printList(mergeTwoLists($l1, $l2));
